cmd/encryption-v1.go

```diff
@@ -168,39 +169,39 @@ func rotateKey(oldKey []byte, newKey []byte, bucket, object string, metadata map
 	}
 }
 
-func newEncryptMetadata(key []byte, bucket, object string, metadata map[string]string, sseS3 bool) ([]byte, error) {
+func newEncryptMetadata(key []byte, bucket, object string, metadata map[string]string, sseS3 bool) (crypto.ObjectKey, error) {
```

Comment:
Normally I would expect this to return `(*crypto.ObjectKey, error)`, so the zero value key wouldn't accidentally get used.

However I suspect that the `nil` value in  some places means "do not use encryption", so returning nil and accidentally using it would have worse side effects.

I wonder if there is a way to avoid this altogether?

How would the zero value key be used, if somebody did so? Will it error?

Reply:
>Normally I would expect this to return `(*crypto.ObjectKey, error)`, so the zero value key wouldn't accidentally get used.

This is something you cannot solve (in Go) to an adequate extent. A caller (not respecting the contract of this function) cannot be stopped doing the wrong thing:
```
key, _ = newEncryptMetadata(...)
...
someCryptoFunction(key, ...)
```
Either it can assign a "default" value it `key` if `key == nil { key = new(crypto.ObjectKey) }` (explicit) or have this implicit in some function:
```
func SsomeCryptoFunction(key *crypto.ObjectKey, ...) {
    if key == nil { ... } // Use default key to not panic in case...  
}
``` 
When writing Go it's part of the contract to check the `error` returned by function call. Not doing so may have severe consequences - deref a nil pointer, not actually creating a file, using a default value where you really shouldn't.

The main reason `crypto.ObjectKey` is a type alias for `[32]byte` compared to `[]byte` is type safety - since a `[]byte` may be nil, empty or have an incorrect or invalid length.


>However I suspect that the nil value in some places means "do not use encryption", so returning nil and accidentally using it would have worse side effects.

That is actually the case. However, all of that should be fixed. A check like `if key == nil` is not an appropriate way to signal that you should not encrypt. 

> How would the zero value key be used, if somebody did so? Will it error?

There is some existing code that treats the zero key as indication of do not **decrypt** (ETag-related) - which is a total hack and should/will be changed. :wink: 

However, that does not mean I'm happy with the shape of the SSE code. There are some parts that are okay-sh - from an API perspective but a lot is too complicated, a hack that needs to be refactored or not very well written - due to e.g. not knowing the overall picture.
Given the resource restrictions (time & priorities) we have I would like to fix all of this (to an acceptable degree) - however, it makes probably more sense to fix it in small chunks then doing to many things in one PR (keep things reviewable and don't (working) break things by accident).

Reply:
It was mostly to determine what was the least harmful side-effect if this was used wrong. Seems the zero value is the most reasonable.